### ISSUE TITLE
libsubprocess: flux_subprocess_read_line() cannot indicate EOF

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#2239]
`flux_subprocess_read_line()` can return a length of 0 if there is not a full line available in the buffer, which means there is no way to detect EOF on the stream through this interface.  The following hack appears in flux-exec.c:
```c
    if (!lenp
        && flux_subprocess_state (p) == FLUX_SUBPROCESS_EXITED) {

        if (!(ptr = flux_subprocess_read (p, stream, -1, &lenp)))
            log_err_exit ("flux_subprocess_read");
    }
```
This is not optimal because there there is no way to detect EOF independent of process termination.

Maybe if the internal buffer has the EOF flag, and there is residual data  in the buffer (not yet terminated as a lline), `flux_subprocess_read_line()` should just return it.  Then, with the change proposed in #2224,  a  len = 0 return from `_read_line()` would signify EOF.

Also is it worth considering changing the prototype of this function:
```c
const char *flux_subprocess_read_line (flux_subprocess_t *p,
                                       const char *stream,
                                       int *lenp);
```
to something that returns a length with -1 indicating error, similar to `read(2)` to be less surprising?  e.g.
```c
int flux_subprocess_read_line (flux_subprocess_t *p,
                               const char *stream,
                               const char **data);
```

**Comment by chu11** [comment]
```C
const char *flux_subprocess_read_line (flux_subprocess_t *p,
                                       const char *stream,
                                       int *lenp);
```

This function returns a `char *` b/c it follows from the `flux_buffer_read_line()` function.  So we can certainly consider changing it, although I would want to consider changing the `flux_buffer_read*()` functions as well.  Perhaps we can consider that in a different issue.  If so, I can open up another issue on it.

**Comment by garlick** [comment]
Agreed, different issue on function signatures.

I have your DONOTMERGE commit in #2246 which does the job.  Shall I add a commit message to it and let it be merged with that PR?

**Comment by chu11** [comment]
I have some `#if 0` and things like that in #2246.  Let me pretty it up just a little bit tomorrow.

**Comment by chu11** [comment]
Pushed a new PR #2254 which is a tad prettied up version of #2245

**Comment by garlick** [comment]
Closed by #2254
